query.go

```diff
@@ -373,11 +373,17 @@ func queryChainServiceBatch(
 				case <-quit:
 					return
 				case <-timeout:
-					if sp.Connected() {
-						continue
-					} else {
-						return
-					}
+
+				// We also check the match signal, since we
+				// might receive an answer for a query that
+				// previously timed out.
+				case <-matchSignal:
```

Comment:
This seems weird to me just from a high level overview. Why would a `matchSignal` be triggered for this peer if it's no longer handling a query? I took a look to see why this would be the case, and it seems to be because we don't modify the `peerStates` map, which tracks the current query a peer is working on. Shouldn't we be clearing a peer's entry within this map once a query times out or matches? Doing this will then require another change further down below when receiving a response through `msgChan`, which is that the current query might not be the one that prompted the response being received.

Reply:
Yeah, you are right, this solution was not bulletproof. See new commit message for amended approach.